### PR TITLE
To dl3

### PR DESCRIPTION
**Pull request by MaxNoe**

### CODE REVIEW

File: klaas/apply.py [hunk 1]
```diff
@@ -43,15 +44,72 @@ def build_query(selection_config):
...
+    generation_config = config.get('feature_generation')
+    print(training_variables)
+    if generation_config:
+        feature_generation(
+            df,
+            generation_config,
+            inplace=True,
+        )
+        training_variables.extend(generation_config['features'])
+    print(training_variables)
```
Comment (mackaiver) [inline klaas/apply.py]: I think these print statements are unnecessary.

File: klaas/scripts/to_dl3.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+@click.argument('output', type=click.Path(exists=False, dir_okay=False))
+@click.option('-k', '--key', help='HDF5 key for h5py hdf5', default='events')
+@click.option('-n', '--n-jobs', type=int, help='Number of cores to use')
+@click.option('-y', '--yes', help='Do not prompt for overwrites', is_flag=True)
+@click.option('-v', '--verbose', help='Verbose log output', is_flag=True)
+@click.option(
+    '-N', '--chunksize', type=int,
+    help='If given, only process the given number of events at once',
+)
+def main(
```
Comment (mackaiver) [inline klaas/scripts/to_dl3.py]: maybe we should rename this script to `to_fact_dl3` or something more specific

File: klaas/scripts/fact_to_dl3.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    try:
+        runs = read_h5py(data_path, key='runs')
+        sources = runs['source'].unique()
+        if len(sources) > 1:
+            raise click.ClickException(
+                'to_dl3 only supports files with a single source'
+            )
+        source = SkyCoord.from_name(sources[0])
+        columns.update(['timestamp', 'night'])
+    except (KeyError, OSError) as e:
```
Comment (mackaiver) [inline klaas/scripts/fact_to_dl3.py]: Why does this catch an OSError?

Comment (MaxNoe) [reply klaas/scripts/fact_to_dl3.py]: From the python docs:

> Changed in version 3.3: EnvironmentError, IOError, WindowsError, socket.error, select.error and mmap.error have been merged into OSError, and the constructor may return a subclass.

**Pull request merged**